migrator/README.md

```diff
@@ -130,19 +130,30 @@ it manipulates and for converting it.
 
 ### Possible migration steps
 
+#### Freeze current schema as needed.
+
+A migration shall not access current schema under `pkg/postgres/schema`. The migration access data in the format
+of its creation time and bring the format of the next migration. It does not evolve with the latest version
+and is associated with a specific sequence number (aka version). The frozen schema helps to keep migration separated
+from Central and keep the codes of migration stable.
+
+For the initial data push to postgres, the generated schemas under `migrator/migrations/frozenschema/v73` were used.
+The auto-generation scripts are removed to freeze the schemas after 3.73.
+
+In 3.74, snapshots of the schemas are taken with an on-demand basis by version. 
+
+Starting from release 4.0, we recommend to keep frozen schemas inside a new migration:
+- If the migration does not change the schema but need to access the data of a table, it needs to freeze its schema in the migration.
```

Comment:
I struggle with this statement.  I'm not sure I understand the why as we need to freeze the schema if we don't need to apply the schema.  Is this for the case if we are accessing the data via the gorm object?  I think what we mainly need to freeze is any store code we are using to access the data here.  I feel like I'm missing something and need clarification on why we need to freeze the schema in this instance.

Reply:
When we access the data, we also use the frozen schema. This is not limited for access using gorm object. If we use the freezed store code, the store codes contains reference to the schema also. It is not likely that the schema is not used at all.

Reply:
that is a good point.  I didn't think the stores needed the schema.  So I took a look and that brings up another potential issue.  It seemed odd to me why the stores in a migration would be using the schema.  We only need the schema for to pass to `search` and we are using `search` for the walk, delete, and get functions in the stores.  Should we be using current search code to access data in an old schema?  Migrations would probably be safer if we did more with sql

Reply:
Is there a policy regarding the content of the `frozenschema` directory hierarchy ?

From the files in `v74`, only `policy_category_edges.go` is new. Looking closer at the schemas and where associated `CreateTable...Stmt` variables are used, I think we should have `v74/policy_category_edges.go` and use it in the `m_170_to_m_171` migration, then `v75/collections.go` and use it in the `m_171_to_m_172` migration.
The other uses of the `v74` schema files should actually rely on the frozen versions of `v73`.
Additionally, the postgres store files should be located in a subdirectory of the appropriate `vXX` schema directory.

The above would allow to have the `m_168_to_m_169` and `m_169_to_m_170` migrations rely on the same frozen schema and store code to remove permissions from permissionsets.

The `m_170_to_m_171` migration would then rely on the `v74` schema and on stores located in various schemas.

I have in mind an upcoming change where I want to enable search on a table, and where the columns for the search would have to be populated from the current content of the proto files (addresses rather the next bullet from the list). I imagine there reading from the `v73` store matching the prior frozen version, and writing back using the `v76` store which created the new columns in the table.

I'll open PRs that puts my ideas in practice and reference them, so that the above ideas become clearer.

Reply:
I tried to materialize my ideas above in the following PRs:
- https://github.com/stackrox/stackrox/pull/4839 (centralize store files associated to schemas and have DDL migration influence frozen schema versions)
- https://github.com/stackrox/stackrox/pull/4840 (postgres -> postgres migration with datamodel update, read from old table schema and write to new table schema)

Reply:
@rhybrillou so your proposal would require a migration to essentially be for a release.  group the store and the schema with the release that changed them.  What about if we make multiple changes within a release that would cause a migration or patch releases?  I was starting to like the idea of keeping these things encapsulated within the specific migration, but I wasn't liking all the potential duplication.  Your approach solves a lot of the duplication issues

Reply:
Yes, exactly. @dashrews78
I will add the duplicate store option in the doc.

Reply:
Sorry for the late answer. My idea was to have one new datamodel version per datamodel change, not per software release. Hence the idea of having `policy_category_edges` added in datamodel `+1`, then `collections` in datamodel `+2` and new columns in `api_tokens` in datamodel `+3`.
I think decoupling the datamodel and migration versions from the software version also simplify the case of multiple changes within a software release. Assuming a table has to undergo multiple schema change in a software release cycle, each migration that changes the datamodel can read in its `old` datamodel and write to its `new` datamodel. The `new` datamodel of a migration can become the `old` datamodel of a later migrations.

Reply:
This has me wondering if it may be simpler to have the datamodel you need within the migration itself.  if we version the schema separately we are now keeping track of software version, database sequence, and now schema version.  Though as I type this I can see some value in versioning the schema separately but if we do so we should figure out a way to do it in a way that is obvious.  Perhaps renaming v73frozenschema to v1frozenschema?

Reply:
The DB sequence number is actually representing all schemas as a whole. We are trying to invent something like GetSchemaAt(table string, seqNum int) and ideally it could pop up a schema(defined in the current master) that fits into the schema at seqNum. It should not be too hard to implement and we would need a special Registry for all frozen schemas. We can get this offline.

Before that happens, we can freeze schema in each migration to be simpler.

Reply:
ROX-15197